File: pkg/auth_test.go
package api

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"testing"

	"github.com/dgrijalva/jwt-go"
)

type roundTripFunc func(*http.Request) (*http.Response, error)

func (f roundTripFunc) RoundTrip(req *http.Request) (*http.Response, error) {
	return f(req)
}

func TestParseAuthResponse(t *testing.T) {
	refreshedRaw := mustRawJWT(t, map[string]any{"email": "user@example.com"})
	response := fmt.Sprintf(`{"tokens":{"skypeToken":"%s","expiresIn":86397},"region":"emea","partition":"emea01","regionGtms":{"chatServiceAggregator":"https://chatsvcagg.teams.microsoft.com"},"regionSettings":{"isUnifiedPresenceEnabled":true,"isOutOfOfficeIntegrationEnabled":true,"isContactMigrationEnabled":true,"isAppsDiscoveryEnabled":true,"isFederationEnabled":true},"licenseDetails":{"isFreemium":false,"isBasicLiveEventsEnabled":true,"isTrial":false,"isAdvComms":false}}`, refreshedRaw)

	client := New(&http.Client{Transport: roundTripFunc(func(req *http.Request) (*http.Response, error) {
		if req.Method != http.MethodPost {
			t.Fatalf("unexpected method: %s", req.Method)
		}
		if req.URL.String() != TEAMS_API_ENDPOINT+"/authsvc/v1.0/authz" {
			t.Fatalf("unexpected URL: %s", req.URL.String())
		}
		if got := req.Header.Get("ms-teams-authz-type"); got != AuthzRefresh {
			t.Fatalf("unexpected authz type: %s", got)
		}
		if got := req.Header.Get("Authorization"); got != "Bearer "+"root-token" {
			t.Fatalf("unexpected authorization header: %s", got)
		}

		return &http.Response{
			StatusCode: http.StatusOK,
			Header:     make(http.Header),
			Body:       io.NopCloser(strings.NewReader(response)),
		}, nil
	})})

	token := &RootSkypeToken{Inner: mustParseJWT(t, "root-token", map[string]any{"email": "user@example.com"}), Type: TokenBearer}
	refreshed, err := client.Authz(token, AuthzRefresh)
	if err != nil {
		t.Fatalf("expected authz to succeed: %v", err)
	}
	if refreshed == nil || refreshed.Inner == nil {
		t.Fatal("expected refreshed token")
	}
	if refreshed.Inner.Raw != refreshedRaw {
		t.Fatalf("unexpected token payload: %s", refreshed.Inner.Raw)
	}
	if refreshed.Type != TokenBearer {
		t.Fatalf("unexpected token type: %s", refreshed.Type)
	}
}

func TestAuthzReturnsTypedError(t *testing.T) {
	client := New(&http.Client{Transport: roundTripFunc(func(req *http.Request) (*http.Response, error) {
		return &http.Response{
			StatusCode: http.StatusUnauthorized,
			Header:     make(http.Header),
			Body:       io.NopCloser(strings.NewReader(`{"errorCode":"GuestUserNotRedeemed","message":"select a tenant first"}`)),
		}, nil
	})})

	_, err := client.Authz(&RootSkypeToken{Inner: mustParseJWT(t, "root-token", map[string]any{"email": "user@example.com"}), Type: TokenBearer}, AuthzRefresh)
	if err == nil {
		t.Fatal("expected authz to fail")
	}
	authzErr, ok := err.(AuthzError)
	if !ok {
		t.Fatalf("expected AuthzError, got %T", err)
	}
	if authzErr.ErrorCode != GuestUserNotRedeemed {
		t.Fatalf("unexpected error code: %s", authzErr.ErrorCode)
	}
}

func TestAuthString(t *testing.T) {
	root := &TeamsToken{Inner: mustParseJWT(t, "root-token", map[string]any{"email": "user@example.com"}), Type: TokenBearer}
	if got := AuthString(root); got != "Bearer "+"root-token" {
		t.Fatalf("unexpected bearer auth string: %s", got)
	}
	root.Type = TokenSkype
	if got := AuthString(root); got != "skypetoken=root-token" {
		t.Fatalf("unexpected skype auth string: %s", got)
	}
	if got := AuthString(nil); got != "" {
		t.Fatalf("unexpected empty auth string: %s", got)
	}
}

func mustParseJWT(t *testing.T, raw string, claims map[string]any) *jwt.Token {
	t.Helper()
	if raw == "" {
		raw = mustRawJWT(t, claims)
	}
	return &jwt.Token{Raw: raw, Claims: jwt.MapClaims(claims)}
}

func mustRawJWT(t *testing.T, claims map[string]any) string {
	t.Helper()
	encodedClaims, err := json.Marshal(claims)
	if err != nil {
		t.Fatalf("unable to marshal claims: %v", err)
	}
	return fmt.Sprintf("%s.%s.signature", base64.RawURLEncoding.EncodeToString([]byte(`{"alg":"none"}`)), base64.RawURLEncoding.EncodeToString(encodedClaims))
}

File: js/tests/auth.test.js
import { describe, expect, test } from 'bun:test';
import { promises as fs } from 'node:fs';
import path from 'node:path';
import { AuthClient, AuthzRefresh } from '../src/auth.js';
import { AuthzError, GenericHTTPError } from '../src/errors.js';
import { authString, getTeamsToken, TeamsToken, TokenBearer, TokenSkype } from '../src/token.js';
import { createFetch, jsonResponse, makeJwt } from './helpers.js';

describe('token helpers', () => {
  test('formats auth strings', () => {
    const token = new TeamsToken(makeJwt({ email: 'user@example.com' }), TokenBearer);
    expect(authString(token)).toBe('Bearer ' + token.inner.raw);
    token.type = TokenSkype;
    expect(authString(token)).toBe(`skypetoken=${token.inner.raw}`);
    expect(authString(null)).toBe('');
  });

  test('loads token from filesystem', async () => {
    const homeDir = await fs.mkdtemp(path.join(process.env.TMPDIR || '/tmp', 'teams-js-'));
    const tokenDir = path.join(homeDir, '.config', 'fossteams');
    await fs.mkdir(tokenDir, { recursive: true });
    const raw = makeJwt({ upn: 'user@example.com' });
    await fs.writeFile(path.join(tokenDir, 'token-teams.jwt'), raw, 'utf8');

    const token = await getTeamsToken({ env: {}, homeDir });
    expect(token.inner.raw).toBe(raw);
  });
});

describe('auth client', () => {
  test('refreshes a token', async () => {
    const refreshed = makeJwt({ email: 'user@example.com' });
    const fetch = createFetch((request) => {
      expect(request.method).toBe('POST');
      expect(request.url).toBe('https://teams.microsoft.com/api/authsvc/v1.0/authz');
      expect(request.headers.get('ms-teams-authz-type')).toBe(AuthzRefresh);
      expect(request.headers.get('Authorization')).toBe('Bearer ' + 'root-token');
      return jsonResponse({ tokens: { skypeToken: refreshed, expiresIn: 3600 } });
    });

    const client = new AuthClient(fetch);
    const token = await client.authz(new TeamsToken({ raw: 'root-token', claims: { email: 'user@example.com' } }, TokenBearer), AuthzRefresh);
    expect(token.inner.raw).toBe(refreshed);
    expect(token.type).toBe(TokenBearer);
  });

  test('returns typed authz errors', async () => {
    const client = new AuthClient(createFetch(() => jsonResponse({ errorCode: 'GuestUserNotRedeemed', message: 'select a tenant first' }, 401)));
    await expect(client.authz(new TeamsToken({ raw: 'root-token', claims: {} }, TokenBearer), AuthzRefresh)).rejects.toBeInstanceOf(AuthzError);
  });

  test('falls back to generic http errors', async () => {
    const client = new AuthClient(createFetch(() => new Response('boom', { status: 500 })));
    await expect(client.authz(new TeamsToken({ raw: 'root-token', claims: {} }, TokenBearer), AuthzRefresh)).rejects.toBeInstanceOf(GenericHTTPError);
  });
});
